### PR TITLE
Serialize NegotiateResponse with IBufferWriter

### PR DESCRIPTION
**Pull request by JamesNK**
https://github.com/aspnet/SignalR/issues/1866

TODO: Remove old implementation

### CODE REVIEW

**Comment by JamesNK** [comment]
```
                           Method |     Mean |     Error |    StdDev |      Op/s |  Gen 0 | Allocated |
--------------------------------- |---------:|----------:|----------:|----------:|-------:|----------:|
       WriteResponse_MemoryStream | 1.715 us | 0.0621 us | 0.0716 us | 582,941.5 | 0.0687 |    6176 B |
 WriteResponse_MemoryBufferWriter | 1.613 us | 0.0169 us | 0.0195 us | 620,031.4 | 0.0038 |     488 B |
```

File: benchmarks/Microsoft.AspNetCore.SignalR.Microbenchmarks/NegotiateProtocolBenchmark.cs [hunk 1]
```diff
@@ -0,0 +1,64 @@
+﻿using System;
```
Comment (mikaelm12) [inline benchmarks/Microsoft.AspNetCore.SignalR.Microbenchmarks/NegotiateProtocolBenchmark.cs]: nit: Header here

File: src/Microsoft.AspNetCore.Http.Connections/HttpConnectionDispatcher.cs [hunk 2]
```diff
@@ -419,7 +420,16 @@ private static byte[] GetNegotiatePayload(string connectionId, HttpContext conte
             var ms = new MemoryStream();
             NegotiateProtocol.WriteResponse(response, ms);
 
-            return ms.ToArray();
+            var writer = MemoryBufferWriter.Get();
+            try
+            {
+                NegotiateProtocol.WriteResponse(response, writer);
+                return writer.ToArray();
```
Comment (JamesNK) [inline src/Microsoft.AspNetCore.Http.Connections/HttpConnectionDispatcher.cs]: I think this `ToArray` is unnessessary. Could pass `MemoryBufferWriter` into this method, serialize response, and then get a `ReadOnlySequence<byte>` from writer and pass that to `context.Response.Body.WriteAsync`.

Just need to add support to `MemoryBufferWriter` for getting a sequence from it.

Comment (JamesNK) [reply src/Microsoft.AspNetCore.Http.Connections/HttpConnectionDispatcher.cs]: Will look into this tomorrow

File: src/Microsoft.AspNetCore.Http.Connections.Common/Microsoft.AspNetCore.Http.Connections.Common.csproj [hunk 3]
```diff
@@ -14,4 +14,7 @@
     <PackageReference Include="Newtonsoft.Json" Version="$(NewtonsoftJsonPackageVersion)" />
     <PackageReference Include="System.Buffers" Version="$(SystemBuffersPackageVersion)" />
   </ItemGroup>
+  <ItemGroup>
+    <ProjectReference Include="..\Microsoft.AspNetCore.SignalR.Common\Microsoft.AspNetCore.SignalR.Common.csproj" />
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Http.Connections.Common/Microsoft.AspNetCore.Http.Connections.Common.csproj]: Nah, this layering is wrong. We need to move the Memory stuff to Common

File: src/Microsoft.AspNetCore.Http.Connections/HttpConnectionDispatcher.cs (hunk 2 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Http.Connections/HttpConnectionDispatcher.cs]: I have some concerns about this and the thread static caching. I don't think it'll play well together.

Comment (davidfowl) [reply src/Microsoft.AspNetCore.Http.Connections/HttpConnectionDispatcher.cs]: Another option would be to not use the thread static cache here but expose the `ReadOnlySequence<byte>` and write it directly response.

**Comment by JamesNK** [comment]
```
                           Method |     Mean |     Error |    StdDev |      Op/s |  Gen 0 | Allocated |
--------------------------------- |---------:|----------:|----------:|----------:|-------:|----------:|
       WriteResponse_MemoryStream | 2.870 us | 0.1807 us | 0.1690 us | 348,425.5 | 0.6866 |    6176 B |
 WriteResponse_MemoryBufferWriter | 1.943 us | 0.0162 us | 0.0151 us | 514,705.6 | 0.0381 |     368 B |
```

image

File: src/Microsoft.AspNetCore.Http.Connections.Common/Microsoft.AspNetCore.Http.Connections.Common.csproj (hunk 3 above)
Comment (davidfowl) [reply src/Microsoft.AspNetCore.Http.Connections.Common/Microsoft.AspNetCore.Http.Connections.Common.csproj]: @JamesNK are you going to move it?

File: src/Microsoft.AspNetCore.Http.Connections/HttpConnectionDispatcher.cs [hunk 4]
```diff
@@ -385,17 +387,27 @@ private Task ProcessNegotiate(HttpContext context, HttpConnectionOptions options
...
+                WriteNegotiatePayload(writer, connection.ConnectionId, context, options);
 
-            // Write it out to the response with the right content length
-            context.Response.ContentLength = negotiateResponseBuffer.Length;
-            return context.Response.Body.WriteAsync(negotiateResponseBuffer, 0, negotiateResponseBuffer.Length);
+                Log.NegotiationRequest(_logger);
+
+                // Write it out to the response with the right content length
+                context.Response.ContentLength = writer.Length;
+                return writer.CopyToAsync(context.Response.Body);
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Http.Connections/HttpConnectionDispatcher.cs]: You need to await here of the MemoryBufferWriter is going to be returned before the operation is complete *if* it goes async.

File: src/Microsoft.AspNetCore.Http.Connections/HttpConnectionManager.cs [hunk 5]
```diff
@@ -207,14 +207,15 @@ public void CloseConnections()
                 // Stop firing the timer
                 _timer?.Dispose();
 
-                var tasks = new List<Task>();
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.Http.Connections/HttpConnectionManager.cs]: Undo, this looks like it snuck in from another change?

File: src/Microsoft.AspNetCore.Http.Connections/HttpConnectionDispatcher.cs (hunk 4 above)
Comment (JamesNK) [reply src/Microsoft.AspNetCore.Http.Connections/HttpConnectionDispatcher.cs]: Good catch

File: src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/HandshakeProtocol.cs [hunk 6]
```diff
@@ -16,6 +17,22 @@ public static class HandshakeProtocol
         private const string ErrorPropertyName = "error";
         private const string TypePropertyName = "type";
 
+        public static readonly ReadOnlyMemory<byte> SuccessHandshakeData;
```
Comment (davidfowl) [inline src/Microsoft.AspNetCore.SignalR.Common/Internal/Protocol/HandshakeProtocol.cs]: Public fields...

**Pull request merged**